src/sentry/snuba/metrics/datasource.py

```diff
@@ -134,6 +140,87 @@ def _get_metrics_by_project_for_entity_query(
+    #
+    # This assumes that metric ids are divided into non-overlapping ranges by the
+    # use case id, so we can focus on a particular range for better performance.
+    min_metric_id = SHARED_STRINGS_PREFIX << 1  # larger than possible metric ids
+    max_metric_id = 0
+
+    for mri, id in SHARED_STRINGS.items():
+        parsed_mri = parse_mri(mri)
+        if parsed_mri is not None and parsed_mri.namespace == use_case_id.value:
+            min_metric_id = min(id, min_metric_id)
+            max_metric_id = max(id, max_metric_id)
+
+    # It's possible that there's a metric id within the use case that is not
+    # hard coded so we should always check the range of custom metric ids.
+    condition = Condition(Column("metric_id"), Op.LT, SHARED_STRINGS_PREFIX)
+
+    # If we find a valid range, we extend the condition to check it as well.
+    if min_metric_id <= max_metric_id:
+        condition = Or(
+            [
+                condition,
+                # Expand the search to include the range of the hard coded
+                # metric ids if a valid range was found.
+                And(
+                    [
+                        Condition(Column("metric_id"), Op.GTE, min_metric_id),
+                        Condition(Column("metric_id"), Op.LTE, max_metric_id),
+                    ]
+                ),
+            ]
+        )
+
+    conditions.append(condition)
+
+    # This is added to every use case id because the MRI is the primary ORDER BY
+    # on the table, and without it, these granules will be scanned no matter what
+    # the use case id is.
+    excluded_mris = []
+
+    if use_case_id == UseCaseID.TRANSACTIONS:
+        # This on_demand MRI takes up the majority of dataset and makes the query slow
+        # because ClickHouse ends up scanning the whole table.
+        #
+        # These are used for on demand metrics extraction and end users should not
+        # need to know about these metrics.
+        #
+        # As an optimization, we explicitly exclude these MRIs in the query to allow
+        # Clickhouse to skip the granules containing strictly these MRIs.
+        if entity_key == EntityKey.GenericMetricsCounters:
+            excluded_mris.append("c:transactions/on_demand@none")
+        elif entity_key == EntityKey.GenericMetricsDistributions:
+            excluded_mris.append("d:transactions/on_demand@none")
+        elif entity_key == EntityKey.GenericMetricsSets:
+            excluded_mris.append("s:transactions/on_demand@none")
+        elif entity_key == EntityKey.GenericMetricsGauges:
+            excluded_mris.append("g:transactions/on_demand@none")
+
+    if excluded_mris:
+        conditions.append(
+            Condition(
+                Column("metric_id"),
+                Op.NOT_IN,
+                # these are shared strings, so just using org id 0 as a placeholder
+                [indexer.resolve(use_case_id, 0, mri) for mri in excluded_mris],
```

Comment:
Can we bulk resolve here? Might slightly improve the performance.

Reply:
Just remembered these are hardcoded mris but I would still parallelize resolution.

Reply:
We don't have a bulk resolve defined but that would not improve performance since this is CPU-bound given that it's just a dict lookup. I agree it would be cleaner if we have a bulk resolve helper, but since we don't, I'm going to leave this as is.

Reply:
Since we don't have it, I am fine with it.